File: src/Localization.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace MsgViewer
{
    public enum AppLanguage
    {
        Korean,
        English,
        French,
        Japanese
    }

    public static class Localization
    {
        public const string FixedLanguageButtonText = "언어 / Language / Langue / 言語";

        public static AppLanguage CurrentLanguage { get; set; }

        static Localization()
        {
            CurrentLanguage = DetectLanguage();
        }

        public static AppLanguage DetectLanguage()
        {
            try
            {
                CultureInfo uiCulture = CultureInfo.CurrentUICulture;
                if (uiCulture == null) return AppLanguage.Korean;

                string name = uiCulture.TwoLetterISOLanguageName.ToLowerInvariant();
                switch (name)
                {
                    case "ko": return AppLanguage.Korean;
                    case "en": return AppLanguage.English;
                    case "fr": return AppLanguage.French;
                    case "ja": return AppLanguage.Japanese;
                    default: return AppLanguage.English;
                }
            }
            catch
            {
                return AppLanguage.Korean;
            }
        }

        public static string AppTitle
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "MSG 뷰어";
                    case AppLanguage.English: return "MSG Viewer";
                    case AppLanguage.French: return "Visualiseur MSG";
                    case AppLanguage.Japanese: return "MSG ビューアー";
                    default: return "MSG Viewer";
                }
            }
        }

        public static string Open
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "열기";
                    case AppLanguage.English: return "Open";
                    case AppLanguage.French: return "Ouvrir";
                    case AppLanguage.Japanese: return "開く";
                    default: return "Open";
                }
            }
        }

        public static string Close
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "닫기";
                    case AppLanguage.English: return "Close";
                    case AppLanguage.French: return "Fermer";
                    case AppLanguage.Japanese: return "閉じる";
                    default: return "Close";
                }
            }
        }

        public static string Exit
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "종료";
                    case AppLanguage.English: return "Exit";
                    case AppLanguage.French: return "Quitter";
                    case AppLanguage.Japanese: return "終了";
                    default: return "Exit";
                }
            }
        }

        public static string From
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "보내는 사람";
                    case AppLanguage.English: return "From";
                    case AppLanguage.French: return "De";
                    case AppLanguage.Japanese: return "差出人";
                    default: return "From";
                }
            }
        }

        public static string SentDate
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "전송";
                    case AppLanguage.English: return "Sent";
                    case AppLanguage.French: return "Envoyé";
                    case AppLanguage.Japanese: return "送信日時";
                    default: return "Sent";
                }
            }
        }

        public static string To
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "받는 사람";
                    case AppLanguage.English: return "To";
                    case AppLanguage.French: return "À";
                    case AppLanguage.Japanese: return "宛先";
                    default: return "To";
                }
            }
        }

        public static string Cc
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "참조";
                    case AppLanguage.English: return "Cc";
                    case AppLanguage.French: return "Cc";
                    case AppLanguage.Japanese: return "㏄";
                    default: return "Cc";
                }
            }
        }

        public static string Bcc
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "숨은참조";
                    case AppLanguage.English: return "Bcc";
                    case AppLanguage.French: return "Cci";
                    case AppLanguage.Japanese: return "Bcc";
                    default: return "Bcc";
                }
            }
        }

        public static string Subject
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "제목";
                    case AppLanguage.English: return "Subject";
                    case AppLanguage.French: return "Objet";
                    case AppLanguage.Japanese: return "件名";
                    default: return "Subject";
                }
            }
        }

        public static string AttachmentsHeader
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "첨부파일";
                    case AppLanguage.English: return "Attachments";
                    case AppLanguage.French: return "Pièces jointes";
                    case AppLanguage.Japanese: return "添付ファイル";
                    default: return "Attachments";
                }
            }
        }

        public static string None
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "없음";
                    case AppLanguage.English: return "None";
                    case AppLanguage.French: return "Aucun";
                    case AppLanguage.Japanese: return "なし";
                    default: return "None";
                }
            }
        }

        public static string NoInformation
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "정보 없음";
                    case AppLanguage.English: return "No information";
                    case AppLanguage.French: return "Aucune information";
                    case AppLanguage.Japanese: return "情報なし";
                    default: return "No information";
                }
            }
        }

        public static string OpenFileDialogTitle
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "MSG 파일 열기";
                    case AppLanguage.English: return "Open MSG File";
                    case AppLanguage.French: return "Ouvrir un fichier MSG";
                    case AppLanguage.Japanese: return "MSGファイルを開く";
                    default: return "Open MSG File";
                }
            }
        }

        public static string MsgFileFilter
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "Outlook 메시지 파일 (*.msg)|*.msg|모든 파일 (*.*)|*.*";
                    case AppLanguage.English: return "Outlook Message Files (*.msg)|*.msg|All Files (*.*)|*.*";
                    case AppLanguage.French: return "Fichiers de message Outlook (*.msg)|*.msg|Tous les fichiers (*.*)|*.*";
                    case AppLanguage.Japanese: return "Outlook メッセージ ファイル (*.msg)|*.msg|すべてのファイル (*.*)|*.*";
                    default: return "Outlook Message Files (*.msg)|*.msg|All Files (*.*)|*.*";
                }
            }
        }

        public static string SaveAttachmentTitle
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "다른 이름으로 저장";
                    case AppLanguage.English: return "Save As";
                    case AppLanguage.French: return "Enregistrer sous";
                    case AppLanguage.Japanese: return "名前を付けて保存";
                    default: return "Save As";
                }
            }
        }

        public static string Error
        {
            get
            {
                switch (CurrentLanguage)
                {
                    case AppLanguage.Korean: return "오류";
                    case AppLanguage.English: return "Error";
                    case AppLanguage.French: return "Erreur";
                    case AppLanguage.Japanese: return "エラー";
                    default: return "Error";
                }
            }
        }

        public static string CannotReadMsgFile(string detail)
        {
            switch (CurrentLanguage)
            {
                case AppLanguage.Korean: return string.Format("MSG 파일을 읽을 수 없습니다:\n{0}", detail);
                case AppLanguage.English: return string.Format("Cannot read MSG file:\n{0}", detail);
                case AppLanguage.French: return string.Format("Impossible de lire le fichier MSG :\n{0}", detail);
                case AppLanguage.Japanese: return string.Format("MSGファイルを読み取れません:\n{0}", detail);
                default: return string.Format("Cannot read MSG file:\n{0}", detail);
            }
        }
    }
}

File: src_native/localization.c
#include "localization.h"

AppLanguage g_currentLanguage = APP_LANG_KO;
const wchar_t* const g_fixedLanguageButtonText = L"언어 / Language / Langue / 言語 (F8)";

static const LocalizationStrings s_stringsKO = {
    L"MSG Viewer (오프라인 휴대용 MSG 뷰어) — v1.2",
    L"열기 (F2)",
    L"닫기 (F4)",
    L"종료",
    L"언어 / Language / Langue / 言語 (F8)",
    L"보내는 사람:",
    L"전송:",
    L"받는 사람:",
    L"참조:",
    L"숨은참조:",
    L"제목:",
    L"첨부파일",
    L"없음",
    L"정보 없음",
    L"MSG 파일 선택",
    L"첨부파일 저장",
    L"Outlook MSG 파일 (*.msg)\0*.msg\0모든 파일 (*.*)\0*.*\0\0",
    L"오류",
    L"1단계: 파일 열기 및 I/O",
    L"2단계: OLE 복합 파일(CFB) 구조 분석",
    L"3단계: MAPI 속성 및 본문 추출",
    L"4단계: 본문 렌더링",
    L"외부 이미지 (오프라인 차단됨)"
};

static const LocalizationStrings s_stringsEN = {
    L"MSG Viewer (Offline Portable MSG Viewer) — v1.2",
    L"Open (F2)",
    L"Close (F4)",
    L"Exit",
    L"언어 / Language / Langue / 言語 (F8)",
    L"From:",
    L"Sent:",
    L"To:",
    L"Cc:",
    L"Bcc:",
    L"Subject:",
    L"Attachments",
    L"None",
    L"No Information",
    L"Select MSG File",
    L"Save Attachment",
    L"Outlook MSG Files (*.msg)\0*.msg\0All Files (*.*)\0*.*\0\0",
    L"Error",
    L"Stage 1: File Open and I/O",
    L"Stage 2: OLE Compound File (CFB) Structure Analysis",
    L"Stage 3: MAPI Property and Body Extraction",
    L"Stage 4: Body Rendering",
    L"Remote Image (Blocked for Offline Protection)"
};

static const LocalizationStrings s_stringsFR = {
    L"MSG Viewer (Visionneuse MSG hors ligne portable) — v1.2",
    L"Ouvrir (F2)",
    L"Fermer (F4)",
    L"Quitter",
    L"언어 / Language / Langue / 言語 (F8)",
    L"De :",
    L"Envoyé :",
    L"À :",
    L"Cc :",
    L"Cci :",
    L"Objet :",
    L"Pièces jointes",
    L"Aucun",
    L"Pas d'information",
    L"Sélectionner un fichier MSG",
    L"Enregistrer la pièce jointe",
    L"Fichiers Outlook MSG (*.msg)\0*.msg\0Tous les fichiers (*.*)\0*.*\0\0",
    L"Erreur",
    L"Étape 1 : Ouverture du fichier et E/S",
    L"Étape 2 : Analyse de la structure du fichier composé OLE (CFB)",
    L"Étape 3 : Extraction des propriétés MAPI et du corps",
    L"Étape 4 : Rendu du corps",
    L"Image distante (Bloquée pour la protection hors ligne)"
};

static const LocalizationStrings s_stringsJA = {
    L"MSG Viewer (オフライン ポータブル MSG ビューアー) — v1.2",
    L"開く (F2)",
    L"閉じる (F4)",
    L"終了",
    L"언어 / Language / Langue / 言語 (F8)",
    L"差出人:",
    L"送信日時:",
    L"宛先:",
    L"CC:",
    L"BCC:",
    L"件名:",
    L"添付ファイル",
    L"なし",
    L"情報なし",
    L"MSG ファイルの選択",
    L"添付ファイルを保存",
    L"Outlook MSG ファイル (*.msg)\0*.msg\0すべてのファイル (*.*)\0*.*\0\0",
    L"エラー",
    L"ステップ 1: ファイルオープンと I/O",
    L"ステップ 2: OLE 複合ファイル (CFB) 構造解析",
    L"ステップ 3: MAPI プロパティおよび本文抽出",
    L"ステップ 4: 本文レンダリング",
    L"リモート画像 (オフライン保護のためブロック)"
};

AppLanguage Localization_DetectLanguage(void) {
    LANGID langId = GetUserDefaultUILanguage();
    WORD primaryLang = PRIMARYLANGID(langId);

    switch (primaryLang) {
        case LANG_KOREAN:   return APP_LANG_KO;
        case LANG_ENGLISH:  return APP_LANG_EN;
        case LANG_FRENCH:   return APP_LANG_FR;
        case LANG_JAPANESE: return APP_LANG_JA;
        default:            return APP_LANG_EN; // Default to English for other locales
    }
}

AppLanguage Localization_GetNextLanguage(AppLanguage current) {
    switch (current) {
        case APP_LANG_KO: return APP_LANG_EN;
        case APP_LANG_EN: return APP_LANG_FR;
        case APP_LANG_FR: return APP_LANG_JA;
        case APP_LANG_JA: return APP_LANG_KO;
        default:          return APP_LANG_KO;
    }
}

const LocalizationStrings* Localization_GetStrings(AppLanguage lang) {
    switch (lang) {
        case APP_LANG_KO: return &s_stringsKO;
        case APP_LANG_EN: return &s_stringsEN;
        case APP_LANG_FR: return &s_stringsFR;
        case APP_LANG_JA: return &s_stringsJA;
        default:          return &s_stringsKO;
    }
}

void Localization_SetLanguage(AppLanguage lang) {
    g_currentLanguage = lang;
}
